File: clientepython/service/ServicioAutor.py
import requests
from datetime import datetime
from typing import List, Optional

from model.Libro import Libro
from model.Autor import Autor

class ServicioAutor:
    BASE_URL = "http://localhost:8080"
    DOMINIO = "/autores"

    def obtener_autores(self) -> List[Autor]:
        response = requests.get(f"{self.BASE_URL}{self.DOMINIO}")
        return [Autor(**autor) for autor in response.json()]

    def obtener_nombres_autores(self) -> List[str]:
        autores = self.obtener_autores()
        return [autor.nombre for autor in autores]

    def obtener_autores_nacionalidad(self, nacionalidad: str) -> List[Autor]:
        response = requests.get(f"{self.BASE_URL}{self.DOMINIO}/nacionalidad/{nacionalidad}")
        return [Autor(**autor) for autor in response.json()]

    def obtener_autores_edad(self, edad: int) -> List[Autor]:
        response = requests.get(f"{self.BASE_URL}{self.DOMINIO}/edad/{edad}")
        return [Autor(**autor) for autor in response.json()]

    def agregar_autor(self, nombre: str, edad: int, nacionalidad: str, 
                      porcentaje_regalias: float, fecha_nacimiento: datetime) -> bool:
        payload = {
            "nombre": nombre,
            "edad": edad,
            "nacionalidad": nacionalidad,
            "regalias": porcentaje_regalias,
            "fecha": fecha_nacimiento
        }
        response = requests.post(
            f"{self.BASE_URL}{self.DOMINIO}/agregar", 
            json=payload, 
            headers={"Content-Type": "application/json"}
        )
        return response.status_code == 200

    def buscar_autor(self, nombre: str) -> Optional[Autor]:
        response = requests.get(f"{self.BASE_URL}{self.DOMINIO}/{nombre}")
        return Autor(**response.json()) if response.status_code == 200 else None

    def actualizar_autor(self, nombre_antiguo: str, nombre: str, edad: int, 
                         nacionalidad: str, porcentaje_regalias: float, 
                         fecha_nacimiento: datetime) -> bool:
        payload = {
            "nombre": nombre,
            "edad": edad,
            "nacionalidad": nacionalidad,
            "regalias": porcentaje_regalias,
            "fecha": fecha_nacimiento
        }
        response = requests.put(
            f"{self.BASE_URL}/autores/actualizar/{nombre_antiguo}", 
            json=payload, 
            headers={"Content-Type": "application/json"}
        )
        return response.status_code == 200

    def eliminar_autor(self, nombre: str) -> bool:
        response = requests.delete(f"{self.BASE_URL}{self.DOMINIO}/borrar/{nombre}")
        return response.status_code == 200

File: ClientePublicacion/Cliente/Cliente/Service/ServicioAutor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Cliente.Model;
using Newtonsoft.Json;
using RestSharp;

namespace Cliente.Service
{
    public class ServicioAutor
    {
        private const string BaseUrl = "http://localhost:8090";
        private const string Dominio = "/autores";

        public ServicioAutor()
        {

        }

        public List<Autor> ObtenerAutores()
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio, Method.Get);

            var response = client.Execute(request);

            return JsonConvert.DeserializeObject<List<Autor>>(response.Content);
        }

        public List<string> ObtenerNombresAutores()
        {
            var autores = ObtenerAutores();

            return autores?.Select(autor => autor.Nombre).ToList() ?? new List<string>();
        }

        public List<Autor> ObtenerAutoresNacionalidad(string nacionalidad)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio + $"/nacionalidad/{nacionalidad}", Method.Get);

            var response = client.Execute(request);

            return JsonConvert.DeserializeObject<List<Autor>>(response.Content);
        }

        public List<Autor> ObtenerAutoresEdad(int edad)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio + $"/edad/{edad}", Method.Get);

            var response = client.Execute(request);

            return JsonConvert.DeserializeObject<List<Autor>>(response.Content);
        }

        public bool AgregarAutor(string nombrex, int edadx, string nacionalidadx, double porcentajeRegaliasx, DateTime fechaNacimientox)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio + "/agregar", Method.Post);


            request.AddHeader("Content-Type", "application/json");

            request.AddBody(new
            {
                nombre = nombrex,
                edad = edadx,
                nacionalidad = nacionalidadx,
                regalias = porcentajeRegaliasx,
                fechaNacimiento = fechaNacimientox.ToString("yyyy-MM-ddTHH:mm:ss")
        });

            var response = client.Execute(request);

            if (response.IsSuccessful)
            {
                return true;
            }
            return false;
        }

        public Autor BuscarAutor(string nombre)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio + $"/buscar/{nombre}", Method.Get);

            var response = client.Execute(request);

            if (response.StatusCode == HttpStatusCode.OK)
            {
                return JsonConvert.DeserializeObject<Autor>(response.Content);
            }
            return null;
        }

        public bool ActualizarAutor(string nombreAntiguo, string nombrex, int edadx, string nacionalidadx, double porcentajeRegaliasx, DateTime fechaNacimientox)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest($"/autores/actualizar/{nombreAntiguo}", Method.Put);

            request.AddHeader("Content-Type", "application/json");

            request.AddJsonBody(new
            {
                nombre = nombrex,
                edad = edadx,
                nacionalidad = nacionalidadx,
                regalias = porcentajeRegaliasx,
                fechaNacimiento = fechaNacimientox.ToString("yyyy-MM-ddTHH:mm:ss")
            });

            var response = client.Execute(request);

            return response.IsSuccessful;
        }

        public bool EliminarAutor(string nombre)
        {
            var options = new RestClientOptions(BaseUrl);
            var client = new RestClient(options);
            var request = new RestRequest(Dominio + $"/borrar/{nombre}", Method.Delete);

            var response = client.Execute(request);

            return response.IsSuccessful;
        }
    }
}
